receiver/dotnetdiagnosticsreceiver/dotnet/parser.go

```diff
@@ -0,0 +1,106 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dotnet
+
+import (
+	"errors"
+	"io"
+
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/dotnetdiagnosticsreceiver/network"
+)
+
+type fieldMsgs map[int]fieldMsg
+
+type Parser struct {
+	r      network.MultiReader
+	mc     RawMetricConsumer
+	logger *zap.Logger
+}
+
+func NewParser(rdr io.Reader, mc RawMetricConsumer, bw network.BlobWriter, logger *zap.Logger) *Parser {
+	pr := network.NewPositionalReader(rdr, bw)
+	r := network.NewMultiReader(pr)
+	return &Parser{r: r, mc: mc, logger: logger}
+}
+
+func (p *Parser) ParseIPC() error {
+	return parseIPC(p.r)
+}
+
+func (p *Parser) ParseNettrace() error {
+	return parseNettrace(p.r)
+}
+
+func (p *Parser) ParseAll() error {
+	var err error
+	fms := fieldMsgs{}
+	for {
+		err = p.parseBlock(fms)
+		// flush regardless of error
+		p.r.Flush()
+		if err != nil {
+			return err
+		}
+	}
+}
+
+func (p *Parser) parseBlock(fms fieldMsgs) error {
+	err := beginPrivateObject(p.r)
+	if err != nil {
+		return err
+	}
+
+	st, err := parseSerializationType(p.r)
+	if err != nil {
+		return err
+	}
+
+	p.logger.Info("parsing block", zap.String("serialization type", st.name))
+
+	switch st.name {
+	case "Trace":
```

Comment:
Maybe I’m just unfamiliar with the code here, but it looks like you’re only parsing one message out of each block. Generally speaking there should be many events per block.

Reply:
Hmm interesting. I can run this receiver against a dotnet v3.1 process on Mac or Linux and it does seem to parse everything, checking the flags at the beginning and end of each object/block (`beginPrivateObject` and `endObject`). I'd imagine it would fail if it weren't parsing all of the messages, because it doesn't skip over content, except for during the 4-byte alignment. There is this ability to parse multiple events/messages in (the admittedly poorly named) `metrics_parser.go`:

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2200/files#diff-0033e0d48a36076e71f7e5fe4f904e415abae5f3c1832234d158831798250ae0R59

and in `metadata_parser.go`:

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2200/files#diff-17259cb98a8dbfdd5966c9ac3c6e3800fde1534bad3d57555bd4fcc0b1d09a07R84

Perhaps this is what you're referring to? If not (and in general 😄), I'm very interested in how to generate messages to cause this parser to fail (I'm currently unable, but I haven't tried very hard, just pointing it at dotnet processes running JetBrains Rider and at a long-running hello world with a very simple, custom EventSource).

Reply:
Ok, I just didn't notice the loops in parseEvent and parseMetadata, and the singular names threw me off. It does look like you're parsing multiple messages out of the blocks.

For testing, the best thing I found to do when I was testing the profiler's EventPipe support was enable every single provider in the runtime and framework and then see if it crashes.

These will cause a ton of events if you set all level/keywords for them:
* Microsoft-Windows-DotNETRuntime
* System.Threading.Tasks.TplEventSource
* System.Runtime
* System.Buffers.ArrayPoolEventSource

And then DiagnosticSource against an ASP.Net app too.

Reply:
Perfect, I will try that (will change the singular names too 😄).